src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorLanguageService`IVsLanguageDebugInfo.cs

```diff
@@ -68,35 +78,27 @@ public int GetProximityExpressions(IVsTextBuffer pBuffer, int iLine, int iCol, i
                 return VSConstants.E_FAIL;
             }
 
-            var dialogResult = _waitDialogFactory.TryCreateWaitDialog(
+            IReadOnlyList<string> proximityExpressions = null;
+            var dialogResult = _uiThreadOperationExecutor.Execute(
                 title: "Determining proximity expressions...",
-                message: "Razor Debugger",
-                async (context) =>
-                {
-                    var proximityExpressions = await _proximityExpressionResolver.TryResolveProximityExpressionsAsync(textBuffer, iLine, iCol, context.CancellationToken).ConfigureAwait(false);
-                    return proximityExpressions;
-                });
-
-            if (dialogResult == null)
-            {
-                // Failed to create the dialog at all.
-                ppEnum = null;
-                return VSConstants.E_FAIL;
-            }
+                defaultDescription: "Razor Debugger",
+                allowCancellation: true,
+                showProgress: true,
+                (context) => _joinableTaskFactory.Run(async () => proximityExpressions = await _proximityExpressionResolver.TryResolveProximityExpressionsAsync(textBuffer, iLine, iCol, context.UserCancellationToken).ConfigureAwait(false)));
```

Comment:
Definitely seems reasonable to enable asynchrony. @jasonmalinowski did you all do something similar?

I'm also a little bit surprised the operation executor doesn't have an async interaction, seems like an API hole by the team? Also feels a bit odd that the method isn't a func (aka ability to return the proximity expression).

I personally loved our API shape better but still better to not maintain the code 😆

Reply:
I think the "standard" way of achieving asynchrony would be [something like what Roslyn did](https://github.com/dotnet/roslyn/pull/52597/files#diff-08d995e2efca001f9f15cf3b4978db893023b97a85c30f85f8b39e2e4a436ab4R54-R64).

Unfortunately the method we're doing this in isn't async (and we can't change it because it's a VS interface), and I'd rather do what we've got right now than what Roslyn did with `SafeContinueWith`.

I agree that our previous API was better for this 😄.

Reply:
> I'm also a little bit surprised the operation executor doesn't have an async interaction, seems like an API hole by the team?

You could file a request for them to add, although you're implicitly opting into whatever JTF wait they'd be doing.

Reply:
Add an extension method? :smile:

Reply:
> Add an extension method? 😄

@ryanbrandenburg I like Jason's idea, an extension method will make this cleaner and more re-usable for the future